src/libraries/System.Private.CoreLib/src/System/Threading/Lock.cs

```diff
@@ -0,0 +1,1194 @@
+                    return true;
+                }
+                return false;
+            }
+
+            private void DecrementWaiterCount()
+            {
+                Debug.Assert(HasAnyWaiters);
+                _state -= WaiterCountIncrement;
+            }
+
+            public bool NeedToSignalWaiter
+            {
+                get
+                {
+                    Debug.Assert(HasAnyWaiters);
+                    return (_state & (SpinnerCountMask | IsWaiterSignaledToWakeMask)) == 0;
+                }
+            }
+
+            public static bool operator ==(State state1, State state2) => state1._state == state2._state;
+            public static bool operator !=(State state1, State state2) => !(state1 == state2);
+
+            bool IEquatable<State>.Equals(State other) => this == other;
+            public override bool Equals(object? obj) => obj is State other && this == other;
+            public override int GetHashCode() => (int)_state;
+
+            private static State CompareExchange(Lock lockObj, State toState, State fromState) =>
+                new State(Interlocked.CompareExchange(ref lockObj._state, toState._state, fromState._state));
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static bool TryLock(Lock lockObj)
+            {
+                // The lock is fair to release waiters in FIFO order, but allows non-waiters to acquire the lock if it's
+                // available to avoid lock convoys.
+                //
+                // Lock convoys can be detrimental to performance in scenarios where work is being done on multiple threads and
+                // the work involves periodically taking a particular lock for a short time to access shared resources. With a
+                // lock convoy, once there is a waiter for the lock (which is not uncommon in such scenarios), a worker thread
+                // would be forced to context-switch on the subsequent attempt to acquire the lock, often long before the worker
+                // thread exhausts its time slice. This process repeats as long as the lock has a waiter, forcing every worker
+                // to context-switch on each attempt to acquire the lock, killing performance and creating a positive feedback
+                // loop that makes it more likely for the lock to have waiters. To avoid the lock convoy, each worker needs to
+                // be allowed to acquire the lock multiple times in sequence despite there being a waiter for the lock in order
+                // to have the worker continue working efficiently during its time slice as long as the lock is not contended.
+                //
+                // This scheme has the possibility to starve waiters. Waiter starvation is mitigated by other means, see
+                // TryLockBeforeSpinLoop() and references to ShouldNotPreemptWaiters.
+
+                var state = new State(lockObj);
```

Comment:
It would be preferable if the fast path does not read the state before trying to lock. The most perf sensitive case of locking is when there is no contention.

Here the first read is just a read - we will fetch and start sharing the cache line and then CompareExchange will request the ownership of the cache line. Since we expect that the lock is not taken, going directly to CompareExchange could be cheaper.

Reply:
> going directly to CompareExchange could be cheaper

Note that his applies only to the very first attempt to lock. Once we learn that the situation is not the trivial/unlocked case, we should do further CompareExchanges conditionally, since they can get expensive. 
It makes sense to do the very first CompareExchange unconditionally though with the expectation that it will succeed.

Reply:
I evaluated both, there was no noticeable difference in perf on x64 or arm64. The code gen was a bit worse than C++, but even then there was no noticeable difference in perf in the uncontended case compared to the previous Lock implementation. I suspect that under the covers the two sets of operations may be fairly similar.

A disadvantage of doing only the 0 to 1 transition in the fast path is that as soon as there is a spinner or waiter registered (fairly common situation), all threads will have to go down the slow path and do at least two CAS operations to acquire the lock.

So based on that I decided to use the same strategy that Coreclr's AwareLock has been using.

Reply:
> as soon as there is a spinner or waiter registered (fairly common situation)

That is the idea of this optimization - the most common situation is when the lock is completely free.  
The next common case is when the lock is used to guard a long-held resource, like a file, and nearly every one waits. In that case extra CompareEchange should not matter.  

The "in-between" case should be generally uncommon. It would require careful balance of time spend inside the lock and outside - to not fall into one of the common cases.  

It could be easier to measure both variants of the initial acquire on the current lock implementation. I'll try to measure. Maybe it makes no difference these days.

Reply:
I have looked at this further and it appears that on recent enough hardware there is no perceivable penalty from reading the shared state prior trying to update it exclusively. 

On the other hand there is advantage with using a modified current state as comparand (vs. simply using "Unocked"), but the reason is not because of spinners. That is because of interaction of waiters and the fast non-spinning path. We allow "unfair" lock grab by the fast path even when there are waiters - to mask the waiter wakeup latency. Such optimization works better when the "unfair" acquire can happen on the first try. That requires that we do not blindly assume "Unlocked" on the fast path.